### ISSUE TITLE
Inconsistent output shape for ONNX model of Ultra-Light-Fast-Generic-Face-Detector-1MB with and without optimization

### ISSUE THREAD

**Issue by tonysung** [Unity-Technologies/barracuda-release#128]
OS: macOS 10.15.6
Barracuda version: 1.2.0-preview
Unity version: 2019.4.17f1

After imported the following ONNX model:

https://github.com/Linzaer/Ultra-Light-Fast-Generic-Face-Detector-1MB/blob/master/models/onnx/version-slim-320.onnx

The inspector shows different output shapes for "scores" with "Optimize Model" turned on and off respectively:

| Off | On |
| - | - |
| <img width="352" alt="image" src="https://user-images.githubusercontent.com/51623/103773994-d4ae2180-5066-11eb-8795-d297d6068186.png"> | <img width="352" alt="image" src="https://user-images.githubusercontent.com/51623/103774031-e7c0f180-5066-11eb-9808-b302dd3128d1.png"> |

In fact, the input shapes are also different (channel first vs channel last).

Cannot test on 1.2.1-preview due to #127

**Comment by AlexRibard** [comment]
@tonysung I have tested your model on latest and validated it against onnx-runtime.
Happy to say that is works :)
I'll notify you once we release the version

**Comment by tonysung** [comment]
Thanks for the swift response @AlexRibard .

If you have a chance, would you also check the inference output w/ the to-be-released version against ONNX runtime as well?

I used this code:

```
float[] floats = new float[320 * 240 * 3];
for (int i = 0; i < floats.Length; ++i) {
    floats[i] = 0.1f;
}
using (var input = new Tensor(1, 240, 320, 3, floats)) {
    worker.Execute(input);
    var scores = worker.PeekOutput("scores");
    yield return new WaitForCompletion(scores);
    Debug.Log(scores[0, 0, 1, 0] + "," + scores[0, 0, 1, 1] + "," + scores[0, 0, 1, 2] + "," + scores[0, 0, 1, 3] + "," + scores[0, 0, 1, 4]);
}
```

Barracuda 1.2.0-preview gives me:
0.5104321,0.4892451,0.5155475,0.4832391,0.5275967

Wrist ONNX python runtime gives me:
[0.10524592 0.10519372 0.1050148  0.10521076 0.10515343]

The latter is more correct.

**Comment by AlexRibard** [comment]
ok, I'll check with your input.
In my test I used generated a random input, fed it to both Barracuda and onnx-runtime and compared the two outputs.
I'll get back to you with your input, but it should be fine

**Comment by AlexRibard** [comment]
confirm that with an input of full of 0.1f I get the same results as onnx-runtime

**Comment by tonysung** [comment]
Thx. Is the update going to be released soon?

**Comment by AlexRibard** [comment]
Not sure, maybe next month?

**Comment by tonysung** [comment]
I'm trying a different (perhaps simpler) model to see if I can get sth to work in the meantime:

https://drive.google.com/file/d/19j7WLsqNEYmX2gUemTlAPQuEwqbYpdM9/view?usp=sharing

I still get different results between onnx runtime and barracuda. Can you please confirm the observation and is it going to be fixed in the upcoming release too?

```
float[] floats = new float[240 * 320 * 3];
for (int i = 0; i < 240 * 320; ++i) {
    floats[i * 3 + 0] = 0.0f;
    floats[i * 3 + 1] = 0.0f;
    floats[i * 3 + 2] = 0.0f;
}

using (var input = new Tensor(1, 240, 320, 3, floats)) {
    worker.Execute(input);
    var scores = worker.PeekOutput("conf");
    yield return new WaitForCompletion(scores);
    Debug.Log(scores[0, 0, 0, 0] + " " + scores[0, 0, 0, 1] + " " + scores[0, 0, 0, 2] + " " + scores[0, 0, 0, 3] + " " + scores[0, 0, 0, 4]);
    Debug.Log(scores[0, 0, 1, 0] + " " + scores[0, 0, 1, 1] + " " + scores[0, 0, 1, 2] + " " + scores[0, 0, 1, 3] + " " + scores[0, 0, 1, 4]);
}

ONNX runtime (Python):
[0.99178046 0.9913564  0.99404293 0.9950336  0.99265504]
[0.0082196  0.00864362 0.00595709 0.00496633 0.00734497]

Barracuda 1.2.0-preview:
0.3997962 0.5737301 0.4188717 0.6445253 0.2490217
0.6002038 0.4262699 0.5811282 0.3554747 0.7509783

Barracuda 1.2.1-preview:
0.3997962 0.5737301 0.4188717 0.6445253 0.2490217
0.6002038 0.4262699 0.5811282 0.3554747 0.7509783
```

**Comment by tonysung** [comment]
I'm trying a different (perhaps simpler) model to see if I can get sth to work in the meantime:

https://drive.google.com/file/d/19j7WLsqNEYmX2gUemTlAPQuEwqbYpdM9/view?usp=sharing

I still get different results between onnx runtime and barracuda. Would you confirm the observation and is it going to be fixed in the upcoming release too?

```
float[] floats = new float[240 * 320 * 3];
for (int i = 0; i < 240 * 320; ++i) {
    floats[i * 3 + 0] = 0.0f;
    floats[i * 3 + 1] = 0.0f;
    floats[i * 3 + 2] = 0.0f;
}

using (var input = new Tensor(1, 240, 320, 3, floats)) {
    worker.Execute(input);
    var scores = worker.PeekOutput("conf");
    yield return new WaitForCompletion(scores);
    Debug.Log(scores[0, 0, 0, 0] + " " + scores[0, 0, 0, 1] + " " + scores[0, 0, 0, 2] + " " + scores[0, 0, 0, 3] + " " + scores[0, 0, 0, 4]);
    Debug.Log(scores[0, 0, 1, 0] + " " + scores[0, 0, 1, 1] + " " + scores[0, 0, 1, 2] + " " + scores[0, 0, 1, 3] + " " + scores[0, 0, 1, 4]);
}

ONNX runtime (Python):
[0.99178046 0.9913564  0.99404293 0.9950336  0.99265504]
[0.0082196  0.00864362 0.00595709 0.00496633 0.00734497]

Barracuda 1.2.0-preview:
0.3997962 0.5737301 0.4188717 0.6445253 0.2490217
0.6002038 0.4262699 0.5811282 0.3554747 0.7509783

Barracuda 1.2.1-preview:
0.3997962 0.5737301 0.4188717 0.6445253 0.2490217
0.6002038 0.4262699 0.5811282 0.3554747 0.7509783
```

**Comment by AlexRibard** [comment]
issue fixed in 1.3.0 cf https://github.com/Unity-Technologies/barracuda-release/issues/127
